src/nvim/message.c

```diff
@@ -890,6 +890,40 @@ char_u *msg_may_trunc(int force, char_u *s)
   return s;
 }
 
+void clear_hl_msg(HlMessage *hl_msg)
+{
+  for (size_t i = 0; i < kv_size(*hl_msg); i++) {
+    xfree(kv_A(*hl_msg, i).text.data);
+  }
+  kv_destroy(*hl_msg);
+  *hl_msg = (HlMessage)KV_INITIAL_VALUE;
+}
+
+#define LINE_BUFFER_SIZE 4096
+
+void add_hl_msg_hist(HlMessage hl_msg)
+{
+  // TODO(notomo): support multi highlighted message history
```

Comment:
I would like to support multi attributes in `MessageHistoryEntry`.
I will try it in next pr because it changes more including `ext_messages` impl.

Reply:
I agree. This PR is useful enough by itself, let's get it merged first.